pkg/geo/wkt/lex.go

```diff
@@ -275,43 +313,164 @@ func (l *wktLex) isValidStrideForLayout(stride int) bool {
 		return stride == 4
 	default:
 		// This should never happen.
-		panic("unknown geom.Layout")
+		panic(fmt.Sprintf("unknown geom.Layout %d", curLayout))
+	}
+}
+
+func (l *wktLex) isNonEmptyAllowedForLayout() bool {
+	if l.getCurLayoutMustBeEmpty() {
+		if l.getCurLayout() != geom.XYM {
+			panic("mustBeEmpty is true but layout is not XYM")
+		}
+		l.setIncorrectLayoutError(geom.NoLayout, "XYM layout must use the M variant of the geometry type")
+		return false
 	}
+	return true
 }
 
 func (l *wktLex) setLayout(layout geom.Layout) bool {
-	if layout == l.curLayout {
+	switch {
+	case layout == l.getCurLayout():
 		return true
-	}
-	if l.curLayout != geom.NoLayout {
+	case l.getCurLayout() != geom.NoLayout:
 		l.setIncorrectLayoutError(layout, "")
 		return false
+	default:
+		l.setLayoutIfNoLayout(layout)
+		return true
 	}
-	l.setLayoutIfNoLayout(layout)
+}
+
+func (l *wktLex) setLayoutBaseType() bool {
```

Comment:
can you add comment to all these setters, and how we should be using them? what is a "base type"? took me a while to get this!